### PR TITLE
Feature - Updating to Head of Alamofire

### PR DESCRIPTION
**Pull request by cnoon**
This PR updates AFI to build against the latest changes in Alamofire. The important callouts are:

* Introduced the new `AFIError` type which supports the two error types that AFI throws directly
* Switched `Request` response serializer extensions over to `DataRequest`. Not going to add support for `DownloadRequest` at this time since it doesn't make tons of sense to do so.
* Response serializers now piggy back on the `serializeDataResponse` API in `Request` to perform the data validation.
* Some of the `AFError` convenience extensions were imported over from the Alamofire tests.

    > Instead of doing this, we could technically import that AF file into our AFI test target and just compile it as part of our test target. Although I don't think this is probably a good idea. Just throwing it out there.

There are no functionality changes anywhere in this PR.

### CODE REVIEW

File: Source/Request+AlamofireImage.swift [hunk 1]
```diff
@@ -159,20 +161,26 @@ extension Request {
...
-            guard error == nil else { return .failure(error!) }
-
-            guard let validData = data, validData.count > 0 else {
-                return .failure(Request.imageDataError())
+    public class func imageResponseSerializer() -> DataResponseSerializer<Image> {
+        return DataResponseSerializer { request, response, data, error in
+            let result = serializeResponseData(response: response, data: data, error: error)
+            let data: Data
+
+            switch result {
```
Comment (jshier) [inline Source/Request+AlamofireImage.swift]: For these types of calls I like to use the `guard` pattern:
```swift
guard case let .sucess(responseData) = result else { return .failure(result.error!) }
```

File: Source/Request+AlamofireImage.swift [hunk 2]
```diff
@@ -202,29 +210,20 @@ extension Request {
 
     // MARK: - Private - Shared Helper Methods
 
-    private class func validateContentType(for request: URLRequest?, response: HTTPURLResponse?) -> Bool {
-        if let url = request?.url, url.isFileURL {
-            return true
-        }
+    private class func validateContentType(for request: URLRequest?, response: HTTPURLResponse?) throws {
```
Comment (jshier) [inline Source/Request+AlamofireImage.swift]: Is there no way to plug into the underlying AF validation code here?

**Comment by jshier** [comment]
@cnoon Review complete, just some minor comments.

File: Source/Request+AlamofireImage.swift (hunk 1 above)
Comment (cnoon) [inline Source/Request+AlamofireImage.swift]: Oh awesome! That's exactly what I was looking for. I didn't realize you could guard that way and expose the `responseData` to the rest of the scope. But it totally makes sense that you can! 

Will change.

File: Source/Request+AlamofireImage.swift (hunk 2 above)
Comment (cnoon) [inline Source/Request+AlamofireImage.swift]: This logic is definitely different, but let me make another pass to see if there's anything that would make sense to pull over.

File: Source/Request+AlamofireImage.swift (hunk 1 above)
Comment (cnoon) [inline Source/Request+AlamofireImage.swift]: Addressed in 1b6aff62.

File: Source/Request+AlamofireImage.swift (hunk 2 above)
Comment (cnoon) [inline Source/Request+AlamofireImage.swift]: Nah, there's really no way to leverage what AF does here. There's a few reasons why. First off, this check is being run in a response serializer, and it probably always needs to be. It's part of validating that the data blob we hand over to `UIImage` for initialization is most likely something that it supports.

We could end up making a `validateImage()` extension on `DataRequest` that would run the same type of validation, but there's no guarantee a user will actually chain that with the `responseImage` serializer.

This is all internal logic and we'd easily be able to make changes to this in a backwards compatible way moving forward. For now, I'd recommend we leave as is and we can revisit later.

**Comment by cnoon** [comment]
All comments have been addressed @jshier...I'm going to move this PR through.

**Pull request merged**